File: server/common/inputProps/formButtonBehavior.go
// This file is part of the Resultra project.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.
package inputProps

import (
	"fmt"
	"github.com/resultra/resultra/server/generic/stringValidation"
	"github.com/resultra/resultra/server/generic/uniqueID"
)

const popupBehaviorModeless string = "modeless"
const popupBehaviorModal string = "modal"

const showFormDestPopup string = "popup"
const showFormDestPage string = "page"
const showFormDestNewPage string = "newPage"

type ButtonPopupBehavior struct {
	PopupMode            string `json:"popupMode"`
	CustomLabelModalSave string `json:"customLabelModalSave"`
	WhereShowForm        string `json:"whereShowForm"`
}

func (srcProps ButtonPopupBehavior) Clone(remappedIDs uniqueID.UniqueIDRemapper) (*ButtonPopupBehavior, error) {
	destProps := srcProps
	return &destProps, nil
}

func NewDefaultPopupBehavior() ButtonPopupBehavior {
	defaultPopupBehavior := ButtonPopupBehavior{
		PopupMode:            popupBehaviorModeless,
		CustomLabelModalSave: "",
		WhereShowForm:        showFormDestNewPage}
	return defaultPopupBehavior
}

func (buttonPopupBehavior ButtonPopupBehavior) ValidateWellFormed() error {

	if !(buttonPopupBehavior.PopupMode == popupBehaviorModeless ||
		buttonPopupBehavior.PopupMode == popupBehaviorModal) {
		return fmt.Errorf("Invalid form popup mode: %v", buttonPopupBehavior.PopupMode)
	}

	if validLabelErr := stringValidation.ValidateOptionalItemLabel(buttonPopupBehavior.CustomLabelModalSave); validLabelErr != nil {
		return validLabelErr
	}

	return nil
}

File: webui/admin/common/inputProperties/formButtonBehavior.js
// This file is part of the Resultra project.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.


function initFormButtonBehaviorProperties(buttonRef,saveBehaviorCallback) {
	
	var $behaviorSelection = $('#formButtonPopupBehaviorSelection')
	var $customModalSaveLabelInput = $('#formButtonCustomModalSaveLabelInput')
	var $whereShowFormSelection = $('#formButtonWhereShowFormSelection')
			
	$behaviorSelection.val(buttonRef.properties.popupBehavior.popupMode)
	$customModalSaveLabelInput.val(buttonRef.properties.popupBehavior.customLabelModalSave)
	$whereShowFormSelection.val(buttonRef.properties.popupBehavior.whereShowForm)
	
	function updateModelLabelVisibility(popupMode) {
		var $modalLabelFormGroup = $('#formButtonCustomModalSaveLabelFormGroup')
		if(popupMode === FormButtonPopupBehaviorModal) {
			$modalLabelFormGroup.show()
		} else {
			$modalLabelFormGroup.hide()
		}	
	}
	
	updateModelLabelVisibility(buttonRef.properties.popupBehavior.popupMode)
	
	function savePropertiesFromControls() {
		var popupBehavior =  {
			popupMode: $behaviorSelection.val(),
			customLabelModalSave: $customModalSaveLabelInput.val(),
			whereShowForm: $whereShowFormSelection.val()
		}
		updateModelLabelVisibility(popupBehavior.popupMode)
		
		saveBehaviorCallback(popupBehavior)		
	}
	
	initSelectControlChangeHandler($behaviorSelection, function() {
		console.log("Popup behavior changed: " + $behaviorSelection.val())
		savePropertiesFromControls()
	})
	initSelectControlChangeHandler($whereShowFormSelection, function() {
		console.log("Popup behavior changed: " + $behaviorSelection.val())
		savePropertiesFromControls()
	})
	
	
	var $popupBehaviorForm = $('#formButtonPopupBehaviorForm')
		
	var remoteValidationParams = {
		url: '/api/generic/stringValidation/validateOptionalItemLabel',
		data: {
			label: function() { return  $customModalSaveLabelInput.val(); }
		}
	}	
			
	var validationSettings = createInlineFormValidationSettings({
		rules: {
			formButtonCustomModalSaveLabelInput: {
				remote: remoteValidationParams
			} // newRoleNameInput
		}
	})	
	
	
	var validator = $popupBehaviorForm.validate(validationSettings)
	
	initInlineInputValidationOnBlur(validator,'#formButtonCustomModalSaveLabelInput',
		remoteValidationParams, function(validatedName) {
			savePropertiesFromControls()	
	})	
	
}